crates/http/src/layers/bytes_to_body_request.rs

```diff
@@ -0,0 +1,71 @@
+// Copyright 2022 The Matrix.org Foundation C.I.C.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use bytes::Bytes;
+use futures_util::future::BoxFuture;
+use http::Request;
+use http_body::Full;
+use tower::{Layer, Service};
+
+#[derive(Clone)]
+pub struct BytesToBodyRequest<S> {
+    inner: S,
+}
+
+impl<S> BytesToBodyRequest<S> {
+    pub const fn new(inner: S) -> Self {
+        Self { inner }
+    }
+}
+
+impl<S> Service<Request<Bytes>> for BytesToBodyRequest<S>
+where
+    S: Service<Request<Full<Bytes>>>,
+    S::Future: Send + 'static,
+{
+    type Error = S::Error;
+    type Response = S::Response;
+    type Future = BoxFuture<'static, Result<Self::Response, Self::Error>>;
+
+    fn poll_ready(
+        &mut self,
+        cx: &mut std::task::Context<'_>,
+    ) -> std::task::Poll<Result<(), Self::Error>> {
+        self.inner.poll_ready(cx)
+    }
+
+    fn call(&mut self, request: Request<Bytes>) -> Self::Future {
+        let (parts, body) = request.into_parts();
+        let body = Full::new(body);
+
+        let request = Request::from_parts(parts, body);
+
+        let inner = self.inner.call(request);
+
+        let fut = async { inner.await };
+
+        Box::pin(fut)
```

Comment:
You probably don't have to box the future here, since we're not mapping anything after.

You could set the associated type `type Future = S::Future;` and then just return `self.inner.call(request)`